doc/how-to-source-packages.rst

```diff
@@ -0,0 +1,309 @@
+   package". That's the ``.tar.gz`` archive containing all of the source for the
+   package.  For example, here's the direct link to download
+   `vector-0.13.1.0.tar.gz <vector-download_>`_.
+
+   Conversely, package *source code* may be only available from a source code
+   repository.  This is the case for all unpublished packages and for all
+   unpublished commits.
+
+.. Warning::
+
+   Be careful with the term "revision". On Hackage this refers to an edited
+   version of the package description. With version control systems, a revision
+   is a commit. Thankfully, when telling Cabal about a "commit" or "revision",
+   the field name is something completely different, a ``tag``.
+
+Fields pointing to *source code* repositories
+---------------------------------------------
+
+Source [code] repositories are a way to specify where to find the source code
+for a package, both for :ref:`package authors<pkg-authors>` and maintainers and
+for :ref:`package consumers<pkg-consumers>`.
+
+A relatively structured set of version control system (VCS) fields, that vary
+depending on the :ref:`VCS kind<vcs-kind>`, enables Cabal commands and other
+tools to interpret and make effective use of this information.
+
+``source-repository``
+   Says where to find the source for a package and is a metadata field of a
+   **package description**. This is the source marker.
+
+``source-repository-package``
+   Locates the source for one or more **project dependencies** that live in the
+   same **source code repository** but in separate folders. The root of the
+   repository is assumed when the ``subdir`` field is omitted.
+
+.. Warning::
+
+   The ``source-repository`` field, the one without "package" is for a package!
+   This actually makes perfect sense as it's within the package description
+   looking outwards.
+
+   .. code-block:: text
+
+       (package description)      source-repository         (Git repository)
+      solver-benchmarks.cabal ──────────────────────────⮚ github/haskell/cabal
+
+      $ cat solver-benchmarks.cabal
+      ...
+      source-repository head
+         type:     git
+         location: https://github.com/haskell/cabal/
+         subdir:   solver-benchmarks
+      ...
+      library
+         build-depends:
+      ...
+            vector
+      ...
+
+   The ``source-repository-package`` field, the one with "package" isn't for a
+   package! It's for a project dependency. This name is a bit misleading because
+   it can describe multiple dependencies and not just one. That it is for
+   "package" also makes sense. Ultimately it is the union of all dependencies of
+   local packages within the project that actually declare the set of project
+   dependencies, not the project.  The project can tighten version constraint
+   ranges with ``constraints`` or loosen them with ``allow-newer`` or
+   ``allow-older`` but it cannot add or remove dependencies.
```

Comment:
This paragraph is hard to follow without an example and a clearer distinction between my local project and the remote/external dependency repository.

Reply:
I made this a separate warning and simplified it.